### PR TITLE
WIP - support molecule v3 and centos8

### PR DESCRIPTION
**Pull request by richm**
This is a big PR
This is based on https://github.com/linux-system-roles/template/pull/39/commits/3add7e85ae97a3a05ccaf79d9a451ca45b0514f0 which fixes the travis docker molecule issue.  That will likely have to be merged and synced sooner rather than later.
commits specific to this PR:
use molecule v3 including new config format - https://github.com/linux-system-roles/template/pull/39/commits/b9b3f5b175491727710744bc5571e427709caa50
remove ansible 2.7 from LSR_ANSIBLES used by molecule - https://github.com/linux-system-roles/template/pull/39/commits/23924df016ce5aba3386364dbd1f7c0f61eaaad8
remove redundant flake8, yamllint checks from molecule - https://github.com/linux-system-roles/template/pull/39/commits/8c0698591d4824a1f86a1a4bc5f74f01b321407a
use stock centos/systemd image for centos7 instead of building and publishing our own - https://github.com/linux-system-roles/template/pull/39/commits/17de4a474c6f0eb5fc23a4a8e4242b993ec3fbd8
add centos8 image to molecule - https://github.com/linux-system-roles/template/pull/39/commits/d224242194cba4f5e5ed28307ca04899383436ba
support podman molecule driver - https://github.com/linux-system-roles/template/pull/39/commits/ac196b354afb15b3ff0b01b343578b30f217a0f3

### CODE REVIEW

**Comment by richm** [comment]
I think it is ok to drop support for ansible 2.7 in molecule. If we need ansible 2.7 testing, we'll have to figure out some other way to do that.

**Comment by richm** [comment]
this pr replaces https://github.com/linux-system-roles/template/pull/11

**Comment by richm** [comment]
would like this to be reviewed and approved by at least two of the core team members

File: molecule/default/molecule.yml
Comment (tyll) [inline molecule/default/molecule.yml:17]: This is removing the container registry, which is bad. See: https://raesene.github.io/blog/2019/09/25/typosquatting-in-a-multi-registry-world/

File: molecule/default/verify.yml
Comment (tyll) [inline molecule/default/verify.yml:1]: There is no license header

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml:17]: ok . . . then we should use quay.io instead of docker.io wherever possible

Comment (richm) [reply molecule/default/molecule.yml]: . . . or registry.centos.org

File: molecule/default/verify.yml
Comment (richm) [reply molecule/default/verify.yml:2]: fixed

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml]: fixed in https://github.com/linux-system-roles/template/pull/39/commits/ab09c6054c3e41306606d4bac3762cc092bff382

File: tox.ini
Comment (tyll) [inline tox.ini]: Why is this needed? Looks to me like `passenv = *` from `[testenv]` should already cover this.

File: molecule/default/molecule.yml
Comment (tyll) [inline molecule/default/molecule.yml:17]: IMHO this should be amended in the first commit to simplify the review.

File: molecule/default/molecule.yml
Comment (tyll) [inline molecule/default/molecule.yml]: Is it correct not to specify a tag here?

File: molecule/default/molecule.yml
Comment (tyll) [inline molecule/default/molecule.yml]: It is strange to add it here without justification and then remove it in the next commit without explanation. This also makes the review more complicated since it requires to review code that is not really used.

File: tox.ini
Comment (tyll) [inline tox.ini:114]: Why is docker needed? The other commit in this PR makes it sound like podman will be used instead.

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml:17]: > IMHO this should be amended in the first commit to simplify the review.

. . . while I thought it would make it easier to review as a followup commit, then squash into a single commit later - but if you would rather have all of the changes made to address review comments in a single commit, ok . . .

File: molecule/default/molecule.yml
Comment (tyll) [inline molecule/default/molecule.yml:6]: The commit message seems to be misleading, it says "use env ..." but it actually adds support for the `LSR_ MOLECULE_DRIVER` variable but does not use it.

File: tox.ini
Comment (tyll) [inline tox.ini:115]: Adding podman does not seem to belong in this commit. Also it seems to depend on an environment variable and does not seem to be needed all the time.

File: tox.ini
Comment (richm) [reply tox.ini]: > Why is this needed? Looks to me like `passenv = *` from `[testenv]` should already cover this.

Looks like it isn't needed - I guess I was confused by all of the `passenv` customization in the other tasks/envs.
So then the question becomes - do we want to restrict the environment for running molecule, in the same manner that we restrict the environment for the other tasks/envs, or do we leave it wide open?

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml]: Apparently the only tag available is `latest` (which is the default if no tag is specified) - https://registry.centos.org/centos/systemd

File: .travis.yml
Comment (tyll) [inline .travis.yml:7]: Not testing Ansible 2.7 should be justified by it being EOL. It could also be in its own PR/commit. Not testing it because molecule does not support it does not seem to be a valid justfication.

**Comment by richm** [comment]
@tyll I'm marking this PR as WIP for now so as not to waste your time with commits attempting to debug the CI failures.  I'll let you know when I have something that I think might meet with your approval

File: .travis.yml
Comment (richm) [reply .travis.yml:7]: `LSR_ANSIBLE` is *only* used for molecule testing

File: tox.ini
Comment (richm) [reply tox.ini:115]: > Adding podman does not seem to belong in this commit. Also it seems to depend on an environment variable and does not seem to be needed all the time.

We still must support the docker runtime for travis, and I suppose many developers may want to use docker by default since podman is relatively new.

How do you suggest writing this such that running in the travis CI environment, and by default, docker is used, but as a developer I can run molecule with the podman driver without having to change any files?

Comment (richm) [reply tox.ini:115]: Sure - installing both `docker` and `podman` isn't "clean" - but what's the harm? And does the cost of doing some really elegant solution justify the benefit?

File: molecule/default/molecule.yml
Comment (tyll) [reply molecule/default/molecule.yml:17]: If you intend to squash the commits, using a fixup commit might be a good idea to indicate this. If the commits are grouped together, they could be easily selected in the commit selection box when reviewing.

In practice I rareley need to see the diff between the original commit and the followup commit since I review the commit by itself. If they are concise  and contain only one change, it is also easy enough to always review them in full.

File: tox.ini
Comment (richm) [reply tox.ini:113]: > Adding podman does not seem to belong in this commit

One of the big reasons for moving to molecule v3 is support for podman.

File: .travis.yml
Comment (richm) [reply .travis.yml:7]: It is required in order to move to molecule v3 - we have to drop support for `LSR_ANSIBLES='ansible==2.7.*'`

Comment (richm) [reply .travis.yml:7]: So to me it logically must be part of the commit to upgrade to molecule v3

Comment (richm) [reply .travis.yml:7]: split this into its own commit

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml:6]: Does this commit clarify your concern?

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml]: I broke up the commits for this PR into discrete but logical units.

**Comment by richm** [comment]
split the fix for the docker travis molecule issue into a separate PR https://github.com/linux-system-roles/template/pull/40 since it is likely we will need to roll this out to the other roles first

**Comment by richm** [comment]
rebased on top of https://github.com/linux-system-roles/template/pull/40
@tyll This PR is ready for your review.

File: tox.ini
Comment (tyll) [inline tox.ini:118]: Should this now specify >= 3?

File: tox.ini
Comment (tyll) [inline tox.ini]: The molecule test envs use the same `envdir`. AFAIU, this will make them all use the same package set so there seems to be no value to specify the dependencies for linting separately. A commit might be nice to explain why they are needed, though.

File: .travis.yml
Comment (tyll) [reply .travis.yml:7]: If we need to support Ansible 2.7 with the system roles, this would be a blocker for adopting molecule V3, I guess. Therefore an explanation why Ansible 2.7 testing can be dropped would be nice in the commit message.

Another note:
Also to keep the repo always in a consistent state, it would be nicer to first apply the commits that prepare molecule v3 support and once this is done, switch to molecule v3. This will keep git bisect useful.

File: molecule/default/molecule.yml
Comment (tyll) [inline molecule/default/molecule.yml]: It would be nicer to remove these before converting them to molecule v3 syntax to reduce the amount of changes.

File: .travis/runtox
Comment (tyll) [inline .travis/runtox:31]: It might be a better idea to first check for podman's existence and then for docker. On recent Fedora, the docker command is an alias for podman and I believe there is no normal docker because it does not work with cgroups v2.

File: molecule/default/molecule.yml
Comment (tyll) [reply molecule/default/molecule.yml:6]: yes, it addresses this concern. But it seems not to default to docker anymore but to autodetect.

File: tox.ini
Comment (tyll) [reply tox.ini]: It is also possible to install podman on the travis Ubuntu machines:
https://travis-ci.community/t/podman-libpod-support/6823/2

> > Adding podman does not seem to belong in this commit
> 
> One of the big reasons for moving to molecule v3 is support for podman.

Since you moved it into its own commit, you addressed this nonetheless.

> I suppose many developers may want to use docker by default since podman is relatively new.

This seems to be worth evaluating to ensure we do not spend unnecessary effort on supporting docker when everyone might be happ(ier) with podman.

Comment (richm) [reply tox.ini]: > This seems to be worth evaluating to ensure we do not spend unnecessary effort on supporting docker when everyone might be happ(ier) with podman.

I would rather do that in a future commit/PR.  I know (anecdotally) that there are a lot of system roles developers who still use `docker`, and I don't want to force them to move to `podman`.  All I really wanted to do in this PR was to give system roles developers the ability to run molecule with either `podman` or `docker`.  So far, the effort to support `docker` is neither unnecessary nor inordinate.

OK?

File: .travis.yml
Comment (richm) [reply .travis.yml:7]: > If we need to support Ansible 2.7 with the system roles, this would be a blocker for adopting molecule V3, I guess. 

Why?  This is only for molecule.

I suppose if it is necessary (as determined by whom?) that we must have molecule support for Ansible 2.7, we can evaluate supporting both molecule v2 *and* molecule v3 in tox/travis.

The way I see it, it isn't necessary that the molecule testing in CI covers all of the Ansible versions we support.  It is another testing tool that gives us some coverage, and any coverage that it doesn't provide, it is up to us to fill in the gaps in molecule coverage via some other method.

>Therefore an explanation why Ansible 2.7 testing can be dropped would be nice in the commit message.

How about what I said above ^^^?

> 
> Another note:
> Also to keep the repo always in a consistent state, it would be nicer to first apply the commits that prepare molecule v3 support and once this is done, switch to molecule v3. This will keep git bisect useful.

I thought I did that.  Can you explain what you mean by "the commits that prepare molecule v3 support", and how the current commits I have can leave the repo in an inconsistent state?  Do you mean "remove redundant flake8, yamllint checks from molecule" which can be applied first?  Can you explain why the current order of commits will cause problems with `git bisect`?

Comment (tyll) [reply .travis.yml:7]: > > If we need to support Ansible 2.7 with the system roles, this would be a blocker for adopting molecule V3, I guess.
> 
> Why? This is only for molecule.
> 
> I suppose if it is necessary (as determined by whom?) that we must have molecule support for Ansible 2.7, we can evaluate supporting both molecule v2 _and_ molecule v3 in tox/travis.
> 
> The way I see it, it isn't necessary that the molecule testing in CI covers all of the Ansible versions we support. It is another testing tool that gives us some coverage, and any coverage that it doesn't provide, it is up to us to fill in the gaps in molecule coverage via some other method.
> 
> > Therefore an explanation why Ansible 2.7 testing can be dropped would be nice in the commit message.
> 
> How about what I said above ^^^?

IMHO the explanation should answer the question "why is it ok not to test Ansible 2.7 with molecule". If this test is optional and already covered by other CI parts, it would be a valid explanation. If we do not need to test Ansible 2.7 at all, because it is EOL, that would be another one. Maybe some documentation why we do molecule at all is missing to ensure that we all know why it is there. To me it appeared to be required not optional and adding some extra checks that we want that are not present elsewhere.

> > Another note:
> > Also to keep the repo always in a consistent state, it would be nicer to first apply the commits that prepare molecule v3 support and once this is done, switch to molecule v3. This will keep git bisect useful.
> 
> I thought I did that. Can you explain what you mean by "the commits that prepare molecule v3 support", and how the current commits I have can leave the repo in an inconsistent state? Do you mean "remove redundant flake8, yamllint checks from molecule" which can be applied first? Can you explain why the current order of commits will cause problems with `git bisect`?

Once the commit "support molecule v3" is applied, the result is broken because molecule v3 support also requires the removal of Ansible 2.7. So if git-bisect would select that commit, tests using molecule would always fail even if there is no change in the role that might have broken something. Therefore it would be nicer to first apply a commit to prepare for molecule v3 support (removing Ansible 2.7) and then do the actual change. Btw. I just realized that the commit is not about "supporting molecule v3" but "switching to molecule v3" since AFAIU, v2 will not be supported after that change.

Comment (richm) [reply .travis.yml:7]: > IMHO the explanation should answer the question "why is it ok not to test Ansible 2.7 with molecule".

Ok, I'll see what I can come up with.

> If we do not need to test Ansible 2.7 at all, because it is EOL, that would be another one.

I think this is the case. @pcahyna ?

The other option is that we support both molecule v2 and v3 until such time as we can make a definitive determination that we will no longer support Ansible 2.7.

> Once the commit "support molecule v3" is applied, the result is broken because molecule v3 support also requires the removal of Ansible 2.7.

What about if I put the change to remove Ansible 2.7 from LSR_ANSIBLES in the same commit as "support molecule v3" (which should be retitled to "switch from molecule v2 to v3")?

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml:6]: > yes, it addresses this concern. But it seems not to default to docker anymore but to autodetect.

If you are running molecule via Travis CI, or running `.travis/runtox` locally, it will autodetect.
If you run `tox -e molecule` you must use `LSR_MOLECULE_DRIVER=podman tox -e molecule` in order to use `podman`.

**Comment by richm** [comment]
Well - looks like we should not drop upstream support for Ansible 2.7 yet - https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html
`2.7 Release Notes | Maintained (security fixes)`
and
```
Ansible maintenance continues for 3 releases. Thus the latest Ansible release receives security and general bug fixes when it is first released, security and critical bug fixes when the next Ansible version is released, and only security fixes once the follow on to that version is released.
```
Since 2.10 is unreleased, 2.7 is still supported for security fixes only.  Therefore, it is a bad idea to drop upstream support for Ansible 2.7.  Note that Ansible 2.7 is EOL downstream but not upstream. https://access.redhat.com/support/policy/updates/ansible-engine - but since this is an upstream project, we should abide by the upstream Ansible lifecycle.

One option would be to put this PR on hold until Ansible 2.7 is officially EOL.

Another option would be to rework this PR so that we have support for molecule v2 and molecule v3 - keep molecule v2 only for Ansible 2.7, and use molecule v3 for later versions.  In travis/tox I suggest we rename the existing molecule to something like "molecule_legacy" and add support for molecule v3 using "molecule" - something like this:
```
[molecule_legacy_common]
changedir = {toxinidir}
deps =
    {env:LSR_ANSIBLE_LEGACY_DEP:ansible}
    docker{env:LSR_MOLECULE_DRIVER_VERSION:}
    molecule<3
    selinux
    wcwidth==0.1.9;python_version=="3.6"
    -rmolecule_extra_requirements.txt
runsyspycmd = {toxinidir}/.travis/runsyspycmd.sh

[testenv:molecule_legacy_version]
changedir = {[molecule_legacy_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule_legacy}
deps =
    {[molecule_legacy_common]deps}
commands =
    bash {[molecule_legacy_common]runsyspycmd} molecule --version
    bash {[molecule_legacy_common]runsyspycmd} ansible --version

[testenv:molecule_legacy_lint]
changedir = {[molecule_legacy_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule_legacy}
deps =
    {[molecule_legacy_common]deps}
commands =
    bash {[molecule_legacy_common]runsyspycmd} \
         molecule lint -s {env:LSR_MSCENARIO:default} {posargs}

[testenv:molecule_legacy_syntax]
changedir = {[molecule_legacy_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule_legacy}
deps =
    {[molecule_legacy_common]deps}
commands =
    bash {[molecule_legacy_common]runsyspycmd} \
         molecule syntax -s {env:LSR_MSCENARIO:default} {posargs}

[testenv:molecule_legacy_test]
changedir = {[molecule_legacy_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule_legacy}
deps =
    {[molecule_legacy_common]deps}
commands =
    bash {[molecule_legacy_common]runsyspycmd} \
         molecule test -s {env:LSR_MSCENARIO:default} {posargs}

[testenv:molecule]_legacy
changedir = {[molecule_legacy_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule_legacy}
deps =
    {[molecule_legacy_common]deps}
commands =
    {[testenv:molecule_legacy_version]commands}
    {[testenv:molecule_legacy_lint]commands}
    {[testenv:molecule_legacy_syntax]commands}
    {[testenv:molecule_legacy_test]commands}

[molecule_common]
changedir = {toxinidir}
deps =
    {env:LSR_ANSIBLE_DEP:ansible}
    ansible-lint
    {env:LSR_MOLECULE_DRIVER:docker}{env:LSR_MOLECULE_DRIVER_VERSION:}
    molecule>=3
    selinux
    wcwidth==0.1.9;python_version=="3.6"
    -rmolecule_extra_requirements.txt
runsyspycmd = {toxinidir}/.travis/runsyspycmd.sh

[testenv:molecule_version]
changedir = {[molecule_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule}
deps =
    {[molecule_common]deps}
commands =
    bash {[molecule_common]runsyspycmd} molecule --version
    bash {[molecule_common]runsyspycmd} ansible --version

[testenv:molecule_lint]
changedir = {[molecule_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule}
deps =
    {[molecule_common]deps}
commands =
    bash {[molecule_common]runsyspycmd} \
         molecule lint -s {env:LSR_MSCENARIO:default} {posargs}

[testenv:molecule_syntax]
changedir = {[molecule_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule}
deps =
    {[molecule_common]deps}
commands =
    bash {[molecule_common]runsyspycmd} \
         molecule syntax -s {env:LSR_MSCENARIO:default} {posargs}

[testenv:molecule_test]
changedir = {[molecule_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule}
deps =
    {[molecule_common]deps}
commands =
    bash {[molecule_common]runsyspycmd} \
         molecule test -s {env:LSR_MSCENARIO:default} {posargs}

[testenv:molecule]
changedir = {[molecule_common]changedir}
envdir = {toxworkdir}/env-{env:TRAVIS_PYTHON_VERSION:molecule}
deps =
    {[molecule_common]deps}
commands =
    {[testenv:molecule_version]commands}
    {[testenv:molecule_lint]commands}
    {[testenv:molecule_syntax]commands}
    {[testenv:molecule_test]commands}
```
Will have to be more involved than that e.g. travis uses the same tox env-3.6 for molecule_legacy and molecule, changes to .travis.yml and .travis/runtox, etc.

**Comment by richm** [comment]
marking as Draft and WIP so we don't accidentally merge it if/until we're ready

**Comment by tyll** [comment]
Ansible 2.10 was released so I guess we can drop Ansible 2.7 support now.

File: tox.ini
Comment (i386x) [reply tox.ini]: IMHO restricting the environment prevents from unintended environment poisoning. OTOH if some environment variables are forgotten to be passed it can have a negative impact on tests.

File: molecule/default/molecule.yml
Comment (i386x) [inline molecule/default/molecule.yml:12]: Just a question: Why we not use `registry.centos.org` also for CentOS 6 image?

**Review by i386x — APPROVED** [review]
lgtm

File: molecule/default/molecule.yml
Comment (richm) [reply molecule/default/molecule.yml:12]: > Just a question: Why we not use `registry.centos.org` also for CentOS 6 image?

I don't know if there is a stock centos6 image with systemd enabled (or for whatever reason we have a custom centos6 image).  That's why I stuck with the linuxsystemroles/centos-6 image.  If we can use a stock image, that would be better.

Comment (tyll) [reply molecule/default/molecule.yml:12]: CentOS 6 is too old for systemd AFAIK.

Comment (richm) [reply molecule/default/molecule.yml:12]: > CentOS 6 is too old for systemd AFAIK.

Then I'm not sure why we need a custom image for centos6 - does anyone know?

**Pull request closed without merging**
